solr/solrj/src/java/org/apache/solr/common/cloud/ZkStateReader.java

```diff
@@ -238,6 +238,55 @@ public class ZkStateReader implements SolrCloseable {
 
   private Future<?> collectionPropsCacheCleaner; // only kept to identify if the cleaner has already been started.
 
+  /**
+   * A ConcurrentHashMap of active watcher by collection name
+   *
+   * Each watcher DocCollectionWatch also contains the latest DocCollection (state) observed
+   */
+  private static class DocCollectionWatches extends ConcurrentHashMap<String, DocCollectionWatch<DocCollectionWatcher>> {
+    /**
+     * Gets the DocCollection (state) of the collection which the corresponding watch last observed
+     * @param collection
+     * @return
+     */
+    private DocCollection getDocCollection(String collection) {
+      DocCollectionWatch<DocCollectionWatcher> watch = get(collection);
+      return watch != null ? watch.currentDoc : null;
+    }
+
+    /**
+     * Updates the latest observed DocCollection (state) of the {@link DocCollectionWatch} if the collection is being watched
+     * @param collection
+     * @param newState  the new DocCollection (state) observed
+     * @return whether an active watch exists for such collection
+     */
+    private boolean updateDocCollection(String collection, DocCollection newState) {
```

Comment:
is this method synchronized somewhere?

Reply:
Sorry I haven't really updated this PR. As I am waiting for the one on apache solr to get approved first, then we will back port the changes from that PR.

In that PR, the update is guarded https://github.com/apache/solr/pull/909/files#diff-c797d60b04b72ceaf84395b8a77c9c51299259d4214e58bcd92331c2bb9c51acR320 😊